File: Code Python/05. Prediction/05.2 Polynomial Regression/PolynomialRegressionHydrodynamics.py
import matplotlib.pyplot as plt
import numpy as np
from functools import reduce

imp = np.genfromtxt('05. Prediction/05.2 Polynomial Regression/yacht_hydrodynamics.csv', delimiter=',')

# Contents:
# 1. Longitudinal position of the center of buoyancy, adimensional.
# 2. Prismatic coefficient, adimensional.
# 3. Length-displacement ratio, adimensional.
# 4. Beam-draught ratio, adimensional.
# 5. Length-beam ratio, adimensional.
# 6. Froude number, adimensional.
# The measured variable is the residuary resistance per unit weight of displacement:
# 7. Residuary resistance per unit weight of displacement, adimensional.

data = np.ones((imp.shape[0], imp.shape[1]))
data[:, 1:] = imp[:, 0:6]
target = imp[:, 6]

traindata = data[1:196, :]
traintarget = target[1:196]

testdata = data[196:, :]
testtarget = target[196:]

# Linear Model

[U, S, V] = np.linalg.svd(traindata, full_matrices=False)
c = reduce(np.dot, [np.matrix.transpose(V), np.linalg.inv(np.diag(S)), np.matrix.transpose(U), traintarget])

plt.figure(figsize=(20, 10))
plt.subplot(2, 3, 1)
plt.plot(testtarget, color='k', linestyle='-', marker='o')
plt.plot(np.dot(testdata, c), color='r', linestyle='-', marker='o')
plt.title('Linear regression model')
plt.legend(['True residuary resistance', 'Predicted residuary resistance'])

ind = np.argsort(testtarget)
plt.subplot(2, 3, 4)
plt.plot(testtarget[ind], color='k', linestyle='-', marker='o')
plt.plot(np.dot(testdata[ind,:], c), color='r', linestyle='-', marker='o')
plt.title('Linear regression model sorted')
plt.legend(['True residuary resistance', 'Predicted residuary resistance'])

# Quadratic Model

data = np.append(data, imp[:, 0:6]**2, axis=1)
traindata = data[1:196, :]
testdata = data[196:, :]

[U, S, V] = np.linalg.svd(traindata, full_matrices=False)
c = reduce(np.dot, [np.matrix.transpose(V), np.linalg.inv(np.diag(S)), np.matrix.transpose(U), traintarget])

plt.subplot(2, 3, 2)
plt.plot(testtarget, color='k', linestyle='-', marker='o')
plt.plot(np.dot(testdata, c), color='r', linestyle='-', marker='o')
plt.title('Quadratic regression model')
plt.legend(['True residuary resistance', 'Predicted residuary resistance'])

ind = np.argsort(testtarget)
plt.subplot(2, 3, 5)
plt.plot(testtarget[ind], color='k', linestyle='-', marker='o')
plt.plot(np.dot(testdata[ind, :], c), color='r', linestyle='-', marker='o')
plt.title('Quadratic regression model sorted')
plt.legend(['True residuary resistance', 'Predicted residuary resistance'])

# Cubic Model

data = np.append(data, imp[:, 0:6]**3, axis=1)
traindata = data[1:196, :]
testdata = data[196:, :]

c = np.linalg.pinv(traindata).dot(traintarget)

plt.subplot(2, 3, 3)
plt.plot(testtarget, color='k', linestyle='-', marker='o')
plt.plot(np.dot(testdata, c), color='r', linestyle='-', marker='o')
plt.title('Cubic regression model')
plt.legend(['True residuary resistance', 'Predicted residuary resistance'])

ind = np.argsort(testtarget)
plt.subplot(2, 3, 6)
plt.plot(testtarget[ind], color='k', linestyle='-', marker='o')
plt.plot(np.dot(testdata[ind,:], c), color='r', linestyle='-', marker='o')
plt.title('Cubic regression model sorted')
plt.legend(['True residuary resistance', 'Predicted residuary resistance'])
plt.show()

File: Code Matlab/05. Prediction/05.2 Polynomial Regression/PolynomialRegressionHydrodynamics.m
clc, clear, close all
%% Import Data
Imp = importdata('yacht_hydrodynamics.dat');

% Contents:
% 1. Longitudinal position of the center of buoyancy, adimensional. 
% 2. Prismatic coefficient, adimensional. 
% 3. Length-displacement ratio, adimensional. 
% 4. Beam-draught ratio, adimensional. 
% 5. Length-beam ratio, adimensional. 
% 6. Froude number, adimensional. 
% The measured variable is the residuary resistance per unit weight of displacement: 
% 7. Residuary resistance per unit weight of displacement, adimensional. 
%% Prepare  Data
data = [ones(size(Imp,1),1),Imp(:,1:6)];
target =  Imp(:,7);

% Predicting data for new airfoil length

traindata =  data(1:196,:);
testdata =  data(197:end,:);

traintarget = target(1:196,:);
testtarget = target(197:end,:);

%% Regression Model linear
[U,S,V] = svd(traindata,'econ');
c = V*inv(S)*U'*traintarget;

figure('Name','Prediction')
subplot(2,3,1)
hold on
plot(testtarget,'k-o');
plot(testdata*c,'r-o');
title('Linear regression model')
legend('True residuary resistance','Predicted residuary resistance',...
    'Location','northwest')


[testtarget_s, ind] = sort(testtarget);
testdata_s = testdata(ind,:);

subplot(2,3,4)
hold on
plot(testtarget_s,'k-o');
plot(testdata_s*c,'r-o');
title('Linear regression model sorted')
legend('True residuary resistance','Predicted residuary resistance',...
    'Location','northwest')


%% Quadratic Regression
data = [data, Imp(:,1:6).^2];

traindata =  data(1:196,:);
testdata =  data(197:end,:);

c = pinv(traindata)*traintarget;


subplot(2,3,2)
hold on
plot(testtarget,'k-o');
plot(testdata*c,'r-o');
title('Quadratic regression model')
legend('True residuary resistance','Predicted residuary resistance',...
    'Location','northwest')

testdata_s = testdata(ind,:);

subplot(2,3,5)
hold on
plot(testtarget_s,'k-o');
plot(testdata_s*c,'r-o');
title('Quadratic regression model sorted')
legend('True residuary resistance','Predicted residuary resistance',...
    'Location','northwest')

%% Cubic Regression

data = [data, Imp(:,1:6).^2, Imp(:,1:6).^3];

traindata =  data(1:196,:);
testdata =  data(197:end,:);

c = pinv(traindata)*traintarget;

subplot(2,3,3)
hold on
plot(testtarget,'k-o');
plot(testdata*c,'r-o');
title('Cubic regression model')
legend('True residuary resistance','Predicted residuary resistance',...
    'Location','northwest')

testdata_s = testdata(ind,:);

subplot(2,3,6)
hold on
plot(testtarget_s,'k-o');
plot(testdata_s*c,'r-o');
title('Cubic regression model sorted')
legend('True residuary resistance','Predicted residuary resistance',...
    'Location','northwest')